src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/QualityScoreDistributionSpark.java

```diff
@@ -0,0 +1,181 @@
+        final JavaRDD<GATKRead> reads = getReads();
+        final JavaRDD<GATKRead> filteredReads = reads.filter(makeReadFilter(PF_READS_ONLY, ALIGNED_READS_ONLY));
+        final Counts result = filteredReads.aggregate(new Counts(INCLUDE_NO_CALLS),
+                (counts, read) -> counts.addRead(read),
+                (counts1, counts2) -> counts1.merge(counts2));
+
+        final MetricsFile<?, Byte> metrics = makeMetrics(result);
+        saveResults(metrics, getHeaderForReads(), getReadSourceName());
+    }
+
+    //Convert the count object into a metrics object so save in a report
+    private MetricsFile<?, Byte> makeMetrics(final Counts result) {
+        // Built the Histograms out of the long[]s
+        final Histogram<Byte> qHisto  = new Histogram<>("QUALITY", "COUNT_OF_Q");
+        final Histogram<Byte> oqHisto = new Histogram<>("QUALITY", "COUNT_OF_OQ");
+
+        for (int i=0; i< result.qCounts.length; ++i) {
+            if (result.qCounts[i]  > 0) qHisto.increment( (byte) i, (double) result.qCounts[i]);
+            if (result.oqCounts[i] > 0) oqHisto.increment((byte) i, (double) result.oqCounts[i]);
+        }
+
+        final MetricsFile<?,Byte> metrics = getMetricsFile();
+        metrics.addHistogram(qHisto);
+        if (!oqHisto.isEmpty()) {
+            metrics.addHistogram(oqHisto);
+        }
+        return metrics;
+    }
+
+    private static Function<GATKRead, Boolean> makeReadFilter(final boolean pfReadOnly, final boolean alignedReadsOnly) {
+        return read -> (!pfReadOnly || !read.failsVendorQualityCheck()) &&
+                (!alignedReadsOnly || !read.isUnmapped()) &&
+                !read.isSecondaryAlignment() &&
+                !read.isSupplementaryAlignment();
+    }
+
+    private void saveResults(final MetricsFile<?, Byte> metrics,  final SAMFileHeader readsHeader, final String inputFileName) {
+        final GCSOptions gcsOptions = getAuthenticatedGCSOptions(); // null if we have no api key
+        try(OutputStream outputStream = BucketUtils.createFile(out, gcsOptions)) {
+            metrics.write(new PrintWriter(outputStream));
+        } catch (final IOException e){
+            throw new GATKException("Could not write metrics to file: " + out, e);
+        }
+        if (metrics.getAllHistograms().isEmpty()) {
+            logger.warn("No valid bases found in input file.");
+        } else if(CHART_OUTPUT != null){
+            // If we're working with a single library, assign that library's name
+            // as a suffix to the plot title
+            String plotSubtitle = "";
+            final List<SAMReadGroupRecord> readGroups = readsHeader.getReadGroups();
+            if (readGroups.size() == 1) {
+                plotSubtitle = readGroups.get(0).getLibrary();
+                if (null == plotSubtitle) {
+                    plotSubtitle = "";
+                }
+            }
+
+            // Now run R to generate a chart
```

Comment:
Add a little more to this comment. Specifically, I was unclear of the relationship between out and CHART_OUTPUT.